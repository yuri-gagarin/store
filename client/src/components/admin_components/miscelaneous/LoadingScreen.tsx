import React from 'react'
import { Dimmer, Loader, Segment } from 'semantic-ui-react'

const LoaderExampleText: React.FC<{}> = (props): JSX.Element => {
  return (
    <div style={{ height: "100vh" }}> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active inverted>
          <Loader>Please Wait</Loader>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default LoaderExampleText