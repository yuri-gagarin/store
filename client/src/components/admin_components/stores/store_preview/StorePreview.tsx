import React from "react";

const StorePreview: React.FC<{}> = (props): JSX.Element => {
  return (
    <div>

    </div>
  );
};

export default StorePreview;