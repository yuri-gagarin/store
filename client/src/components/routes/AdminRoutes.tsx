import React from "react";
import { Route } from "react-router-dom";
// component imports //
import LoginComponent from "../admin_components/login/LoginComponent";

const AdminRoutes: React.FC<{}> = (props): JSX.Element =>  {
  return (
    <React.Fragment>
      <Route exact path="/admin/login">
        <LoginComponent />
      </Route>
      
    </React.Fragment>
    
  );
};

export default AdminRoutes;