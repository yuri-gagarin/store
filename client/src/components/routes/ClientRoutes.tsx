import React from "react";
import { Switch, Route } from "react-router-dom";
// component imports //
import HomeScreenComponent from "../client_components/home_screen/HomeScreenComponent";
import StoreHolder from "../client_components/store_screen/StoreHolder";
import MainNavSideBar from "../client_components/navbars/MainNavSidebar";

const ClientRoutes: React.FC<{}> = (props): JSX.Element =>  {
  return (
    <MainNavSideBar>
      <Switch>
        <Route exact path="/">
          <HomeScreenComponent title=""/>
        </Route>
        <Route exact path="/store">
          <StoreHolder />
        </Route>
      </Switch>
    </MainNavSideBar>
    
  );
};

export default ClientRoutes;