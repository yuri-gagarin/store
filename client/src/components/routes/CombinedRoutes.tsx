import React from "react";
import { BrowserRouter, Route, Switch } from "react-router-dom";
import MainNavSideBar from "../client_components/navbars/MainNavSidebar";
import LoginComponent from "../admin_components/login/LoginComponent";
import AdminHomeScreen from "../admin_components/home/AdminHomeScreen";
// import AdminRoutes from "./AdminRoutes";
// component imports //

const CombinedRoutes: React.FC<{}> = (props): JSX.Element =>  {
  return (
    <BrowserRouter>
      <Switch>
        <Route path="/admin/login">
          <LoginComponent />
        </Route>
        <Route path="/admin/home">
          <AdminHomeScreen />
        </Route>
        <Route path="/">
          <MainNavSideBar />
        </Route>
      </Switch>
      
      
    </BrowserRouter>
  );
};

export default CombinedRoutes;