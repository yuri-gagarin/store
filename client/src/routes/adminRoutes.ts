export const AdminProductRoutes = {
  HOME_ROUTE: "/admin/home/my_products",
  VIEW_ALL_ROUTE: "/admin/home/my_products/view_all",
  CREATE_ROUTE: "/admin/home/my_products/create",
  MANAGE_ROUTE: "/admin/home/my_products/manage",
  SHOW_ROUTE: "/admin/home/my_products/show",
  VIEW_SORTED_ROUTE: "/admin/home/my_products/view_sorted"
};

export const AdminServiceRoutes = {
  ADMIN_SERVICES_HOME_ROUTE: "/admin/home/my_services",
  ADMIN_SERVICES_VIEW_ALL_ROUTE: "/admin/home/my_services/view_all",
  ADMIN_SERVICES_CREATE_ROUTE: "/admin/home/my_services/create",
  ADMIN_SERVICES_MANAGE_ROUTE: "/admin/home/my_services/manage",
  ADMIN_SERVICES_SHOW_ROUTE: "/admin/home/my_services/show",
  ADMIN_SERVICES_VIEW_SORTED_ROUTE: "/admin/home/my_services/view_sorted"
};

export const AdminStoreRoutes = {
  ADMIN_STORES_HOME_ROUTE: "/admin/home/my_stores",
  ADMIN_STORES_CREATE_ROUTE: "/admin/home/my_stores/create",
  ADMIN_STORES_MANAGE_ROUTE: "/admin/home/my_stores/manage",
  ADMIN_STORES_SHOW_ROUTE: "/admin/home/my_stores/show",
  ADMIN_STORES_VIEW_ALL_ROUTE: "/admin/home/my_stores/view_all"
};

export const AdminStoreItemRoutes = {
  HOME_ROUTE: "/admin/home/my_store_items",
  CREATE_ROUTE: "/admin/home/my_store_items/create",
  MANAGE_ROUTE: "/admin/home/my_store_items/manage",
  SHOW_ROUTE: "/admin/home/my_store_items/show",
  VIEW_ALL_ROUTE: "/admin/home/my_store_items/view_all",
  VIEW_SORTED_ROUTE: "/admin/home/my_store_items/view_sorted"
};

