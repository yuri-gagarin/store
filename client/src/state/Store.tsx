import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

