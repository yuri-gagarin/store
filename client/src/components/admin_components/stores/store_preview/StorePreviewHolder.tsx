import React from "react";
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// additional components //
import StorePreview from "./StorePreview";
import LoadingScreen from "../../miscelaneous/LoadingScreen";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>
}

const StorePreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { loadedStores, loading } = state.storeState;
  if (loading) {
    return (
      <LoadingScreen />
    );
  } else {
    return (
      <div>
        {
          loadedStores.map((store) => {
            return (
              <StorePreview />
            )
          })
        }
      </div>
    );
  }
};

export default StorePreviewHolder;