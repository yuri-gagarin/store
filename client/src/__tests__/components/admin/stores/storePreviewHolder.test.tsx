import React from "react";
import { MemoryRouter as Router } from "react-router-dom";
// test dependencies //
import moxios from "moxios";
import { mount, ReactWrapper } from "enzyme"; 
import { act } from "react-dom/test-utils";
// components //
import StorePreviewHolder from "../../../../components/admin_components/stores/store_preview/StorePreviewHolder";
import StorePreview from "../../../../components/admin_components/stores/store_preview/StorePreview";
import LoadingScreen from "../../../../components/admin_components/miscelaneous/LoadingScreen";
// state //
import { TestStateProvider } from "../../../../state/Store";
// helpers //
import { generateCleanState } from "../../../../test_helpers/miscHelpers";
import { createMockStores } from "../../../../test_helpers/storeHelpers";
// admin routes //
import { AdminStoreRoutes } from "../../../../routes/adminRoutes";

describe("StorePreviewHolder Component render tests", () => {
  // TEST StprePreviewHolder in its loading state //
  describe("StorePreviewHolder in 'loading' state", () => {
    let wrapper: ReactWrapper;
    beforeAll(() => {
      const mockState = generateCleanState();
      mockState.storeState.loading = true;
      wrapper = mount(
        <TestStateProvider mockState={mockState}>
          <Router keyLength={0} initialEntries={[ AdminStoreRoutes.VIEW_ALL_ROUTE  ]}>
            <StorePreviewHolder />
          </Router>
        </TestStateProvider>
      );
    });
    it("Should correctly render", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
    it("Should have a LoadingScreen while 'loading == true'", () => {
      const loadingScreen = wrapper.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1);
    });
    it("Should not display the '#adminStorePreviewHolder'", () => {
      const adminStorePreview = wrapper.find("#adminStorePreviewHolder");
      expect(adminStorePreview.length).toEqual(0);
    }); 
    it("Should not display any StorePreview Components", () => {
      const storePreviewComponents = wrapper.find(StorePreview);
      expect(storePreviewComponents.length).toEqual(0);
    });

  });
  // END TEST StorePreviewHolder in its loading state //
  // TEST StorePreviewHolder in its loaded state //
  describe("StorePreview in 'loaded' state", () => {
    let wrapper: ReactWrapper; let mockStores: IStoreData[];
    beforeAll( async () => {
      moxios.install();
      const mockState = generateCleanState();
      // set mock stores //
      mockStores = [
        {
          _id: "1",
          title: "first",
          description: "desc",
          images: [],
          createdAt: "now"
        },
        {
          _id: "2",
          title: "second",
          description: "desc",
          images: [],
          createdAt: "now"
        },
        {
          _id: "3",
          title: "third",
          description: "desc",
          images: [],
          createdAt: "now"
        }
      ];

      wrapper = mount(
         <TestStateProvider mockState={mockState}>
           <Router keyLength={0} initialEntries={[AdminStoreRoutes.VIEW_ALL_ROUTE]}>
             <StorePreviewHolder />
           </Router>
         </TestStateProvider>
      );

      await act( async () => {
        await moxios.stubRequest("/api/stores", {
          status: 200,
          response: {
            responseMsg: "All ok",
            stores: mockStores
          }
        });
      });
      wrapper.update();
    });
    it("Should correctly render", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
    it("Should NOT have a LoadingScreen while 'loading == false'", () => {
      const loadingScreen = wrapper.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(0);
    });
    it("Should display the '#adminStorePreviewHolder'", () => {
      const adminStorePreview = wrapper.find("#adminStorePreviewHolder");
      expect(adminStorePreview.length).toEqual(1);
    });
    it(`Should display a correct (3) number of StorePreview Components`, () => {
      const storePreviewComponents = wrapper.find(StorePreview);
      expect(storePreviewComponents.length).toEqual(mockStores.length);
    });
  });
  // END TEST StorePreviewHolder in its loaded state //
});