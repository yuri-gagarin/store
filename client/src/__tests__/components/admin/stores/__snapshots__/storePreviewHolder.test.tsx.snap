// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorePreviewHolder Component render tests StorePreview in 'loaded' state Should correctly render 1`] = `
<div
  id="adminStorePreviewHolder"
>
  <div
    class="ui segment"
  >
    <h3
      class="adminStoreTitle"
    >
      first
    </h3>
    <div
      class="adminStoreDescription"
    >
      desc
    </div>
    <ul
      class="adminStoreDetails"
    >
      <li>
        Created At: NaN-undefined-NaN
      </li>
      <li>
        Edited At: Not Available
      </li>
    </ul>
    <div
      class="adminStoreImageDetails"
    >
      <h5>
        Displayed Images: 
      </h5>
    </div>
    <div
      aria-live="polite"
      class="image-gallery"
    >
      <div
        class="image-gallery-content bottom"
      >
        <div
          class="image-gallery-slide-wrapper bottom"
        >
          <div
            class="image-gallery-slides"
          />
          <button
            aria-label="Play or Pause Slideshow"
            class="image-gallery-icon image-gallery-play-button"
            type="button"
          >
            <svg
              class="image-gallery-svg"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="5 3 19 12 5 21 5 3"
              />
            </svg>
          </button>
          <button
            aria-label="Open Fullscreen"
            class="image-gallery-icon image-gallery-fullscreen-button"
            type="button"
          >
            <svg
              class="image-gallery-svg"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 3H5a2 2 0 0 0-2 2v3m18 0V5a2 2 0 0 0-2-2h-3m0 18h3a2 2 0 0 0 2-2v-3M3 16v3a2 2 0 0 0 2 2h3"
              />
            </svg>
          </button>
        </div>
        <div
          class="image-gallery-thumbnails-wrapper bottom"
        >
          <div
            class="image-gallery-thumbnails"
          >
            <div
              aria-label="Thumbnail Navigation"
              class="image-gallery-thumbnails-container"
              style="transform: translate3d(0px, 0, 0);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ui segment"
  >
    <h3
      class="adminStoreTitle"
    >
      second
    </h3>
    <div
      class="adminStoreDescription"
    >
      desc
    </div>
    <ul
      class="adminStoreDetails"
    >
      <li>
        Created At: NaN-undefined-NaN
      </li>
      <li>
        Edited At: Not Available
      </li>
    </ul>
    <div
      class="adminStoreImageDetails"
    >
      <h5>
        Displayed Images: 
      </h5>
    </div>
    <div
      aria-live="polite"
      class="image-gallery"
    >
      <div
        class="image-gallery-content bottom"
      >
        <div
          class="image-gallery-slide-wrapper bottom"
        >
          <div
            class="image-gallery-slides"
          />
          <button
            aria-label="Play or Pause Slideshow"
            class="image-gallery-icon image-gallery-play-button"
            type="button"
          >
            <svg
              class="image-gallery-svg"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="5 3 19 12 5 21 5 3"
              />
            </svg>
          </button>
          <button
            aria-label="Open Fullscreen"
            class="image-gallery-icon image-gallery-fullscreen-button"
            type="button"
          >
            <svg
              class="image-gallery-svg"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 3H5a2 2 0 0 0-2 2v3m18 0V5a2 2 0 0 0-2-2h-3m0 18h3a2 2 0 0 0 2-2v-3M3 16v3a2 2 0 0 0 2 2h3"
              />
            </svg>
          </button>
        </div>
        <div
          class="image-gallery-thumbnails-wrapper bottom"
        >
          <div
            class="image-gallery-thumbnails"
          >
            <div
              aria-label="Thumbnail Navigation"
              class="image-gallery-thumbnails-container"
              style="transform: translate3d(0px, 0, 0);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ui segment"
  >
    <h3
      class="adminStoreTitle"
    >
      third
    </h3>
    <div
      class="adminStoreDescription"
    >
      desc
    </div>
    <ul
      class="adminStoreDetails"
    >
      <li>
        Created At: NaN-undefined-NaN
      </li>
      <li>
        Edited At: Not Available
      </li>
    </ul>
    <div
      class="adminStoreImageDetails"
    >
      <h5>
        Displayed Images: 
      </h5>
    </div>
    <div
      aria-live="polite"
      class="image-gallery"
    >
      <div
        class="image-gallery-content bottom"
      >
        <div
          class="image-gallery-slide-wrapper bottom"
        >
          <div
            class="image-gallery-slides"
          />
          <button
            aria-label="Play or Pause Slideshow"
            class="image-gallery-icon image-gallery-play-button"
            type="button"
          >
            <svg
              class="image-gallery-svg"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="5 3 19 12 5 21 5 3"
              />
            </svg>
          </button>
          <button
            aria-label="Open Fullscreen"
            class="image-gallery-icon image-gallery-fullscreen-button"
            type="button"
          >
            <svg
              class="image-gallery-svg"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 3H5a2 2 0 0 0-2 2v3m18 0V5a2 2 0 0 0-2-2h-3m0 18h3a2 2 0 0 0 2-2v-3M3 16v3a2 2 0 0 0 2 2h3"
              />
            </svg>
          </button>
        </div>
        <div
          class="image-gallery-thumbnails-wrapper bottom"
        >
          <div
            class="image-gallery-thumbnails"
          >
            <div
              aria-label="Thumbnail Navigation"
              class="image-gallery-thumbnails-container"
              style="transform: translate3d(0px, 0, 0);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`StorePreviewHolder Component render tests StorePreviewHolder in 'loading' state Should correctly render 1`] = `
<div
  id="generalLoadingScreen"
  style="height: 100vh;"
>
  <div
    class="ui segment"
    style="height: 100vh;"
  >
    <div
      class="ui active transition visible inverted dimmer"
      style="display: flex;"
    >
      <div
        class="content"
      >
        <div
          class="ui text loader"
        >
          Please Wait
        </div>
      </div>
    </div>
  </div>
</div>
`;
